File: apps/driver-android/src/main/kotlin/lk/mageride/driver/history/RatePassengerSheet.kt
package lk.mageride.driver.history

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.style.TextAlign
import lk.mageride.driver.R
import lk.mageride.driver.jobs.ScheduleLabels
import lk.mageride.driver.ui.Symbols
import lk.mageride.driver.ui.component.LabelledTextField
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme

/**
 * **The rate-passenger sheet** — AL-35's ruling, drawn.
 *
 * > *"Tapping 'Rate ★' on a completed trip opens a **modal bottom sheet** to rate the passenger 1–5
 * > stars (+ optional comment) (US-18.2) — it is **no longer an inline card**."*
 *
 * The wireframe's contents, top to bottom: the grab handle, *"Rate passenger"*, the passenger's
 * avatar with their name and the trip under it, the five stars, the optional comment field and the
 * `Submit rating` CTA.
 *
 * **One completed ride can never be rated: a proxy booking whose rider is unregistered** (P-01).
 * There is no `iam.users` row to be the `ratee_id`, and `DriverRatingInput.passengerId` is required
 * — so the sheet says so and the CTA stays dead rather than posting a rating about nobody.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun RatePassengerSheet(
    state: RatePassengerState?,
    viewModel: RideHistoryViewModel,
    modifier: Modifier = Modifier,
) {
    if (state == null) return
    val locale = LocalConfiguration.current.locales[0]

    ModalBottomSheet(
        onDismissRequest = viewModel::dismissRating,
        modifier = modifier,
        sheetState = rememberModalBottomSheetState(),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start = MageRideTheme.spacing.md,
                    end = MageRideTheme.spacing.md,
                    bottom = MageRideTheme.spacing.lg,
                ),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            Text(
                text = stringResource(R.string.history_rate_title),
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )

            PassengerRow(
                name = state.passengerName ?: stringResource(R.string.history_passenger_unnamed),
                trip = "${tripRoute(state.trip.summary)} ${Symbols.DOT} " +
                    ScheduleLabels.date(state.trip.summary.startedAt, locale),
            )

            StarRow(stars = state.stars, onStarsChange = viewModel::onStarsChange)

            LabelledTextField(
                label = stringResource(R.string.history_rate_comment_label),
                value = state.comment,
                onValueChange = viewModel::onCommentChange,
                placeholder = stringResource(R.string.history_rate_comment_hint),
                singleLine = false,
            )

            if (!state.loading && state.passengerId == null) {
                Text(
                    text = stringResource(R.string.history_rate_no_passenger),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            MageRideCta(
                label = stringResource(R.string.history_rate_submit),
                onClick = viewModel::submitRating,
                enabled = state.canSubmit,
                loading = state.submitting,
            )
        }
    }
}

/** The wireframe's avatar and the two lines beside it. */
@Composable
private fun PassengerRow(name: String, trip: String, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Surface(
            modifier = Modifier.size(ControlTokens.AvatarSmall),
            shape = CircleShape,
            color = MaterialTheme.colorScheme.primaryContainer,
        ) {
            Box(contentAlignment = Alignment.Center) {
                Icon(
                    imageVector = Icons.Outlined.Person,
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.RowIcon),
                    tint = MaterialTheme.colorScheme.onPrimaryContainer,
                )
            }
        }
        Column {
            Text(
                text = name,
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
            Text(
                text = trip,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

/**
 * The five stars, as five tap targets.
 *
 * `role = Role.RadioButton` on each: one of five is exactly what a star rating is, and TalkBack
 * announcing five buttons would leave a driver with no way to know which was chosen.
 */
@Composable
private fun StarRow(stars: Int, onStarsChange: (Int) -> Unit, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs, Alignment.CenterHorizontally),
    ) {
        for (star in RatePassengerState.MIN_STARS..RatePassengerState.MAX_STARS) {
            Text(
                text = if (star <= stars) Symbols.STAR_FILLED else Symbols.STAR_EMPTY,
                modifier = Modifier
                    .clickable(
                        role = Role.RadioButton,
                        onClickLabel = star.toString(),
                        onClick = { onStarsChange(star) },
                    )
                    .padding(MageRideTheme.spacing.xxs),
                style = MaterialTheme.typography.headlineMedium,
                color = MageRideTheme.status.warning,
                textAlign = TextAlign.Center,
            )
        }
    }
}

File: apps/driver-ios/DriverApp/History/RatePassengerSheet.swift
import MageRideShared
import SwiftUI

/// **The rate-passenger sheet** — AL-35's ruling, drawn.
///
/// > *"Tapping 'Rate ★' on a completed trip opens a sheet (iOS `.sheet`, detent `.medium`) to rate the
/// > passenger 1–5 stars (+ optional comment) (US-18.2) — **no longer an inline card**."*
///
/// The wireframe's contents, top to bottom: the grab handle (which a presented `.sheet` draws itself),
/// *"Rate passenger"*, the passenger's avatar with their name and the trip under it, the five stars,
/// the optional comment field and the `Submit rating` CTA.
///
/// **One completed ride can never be rated: a proxy booking whose rider is unregistered** (P-01). There
/// is no `iam.users` row to be the `ratee_id`, and `DriverRatingInput.passengerId` is required — so the
/// sheet says so and the CTA stays dead rather than posting a rating about nobody.
///
/// **The sheet reads its own state off the model rather than being handed a value**, because the
/// passenger's name arrives from `GET /v1/rides/{rideId}` a moment *after* the sheet opens — see
/// ``RideHistoryScreen`` for why that rules out `sheet(item:)`.
@MainActor
struct RatePassengerSheet: View {

    @ObservedObject var model: RideHistoryModel

    var body: some View {
        if let rating = model.state.rating {
            sheet(rating)
        }
    }

    private func sheet(_ rating: RatePassengerState) -> some View {
        NavigationStack {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                passenger(rating)

                StarRow(
                    stars: rating.stars,
                    onStarsChange: model.onStarsChange
                )

                LabelledTextField(
                    labelKey: "history_rate_comment_label",
                    value: Binding(get: { rating.comment }, set: model.onCommentChange),
                    placeholder: "history_rate_comment_hint".localised,
                    autocapitalisation: .sentences
                )

                if !rating.isLoading && rating.passengerId == nil {
                    Text(key: "history_rate_no_passenger")
                        .mageFont(.label)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }

                if let errorKey = model.state.errorKey {
                    FormErrorText(messageKey: errorKey)
                }

                Button {
                    Task { await model.submitRating() }
                } label: {
                    Text(key: "history_rate_submit")
                }
                .buttonStyle(.mageCta(loading: rating.isSubmitting))
                .disabled(!rating.canSubmit)

                Spacer(minLength: 0)
            }
            .padding(MageRideSpacing.md)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(MageRideColor.background)
            .navigationTitle(Text(key: "history_rate_title"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(action: model.dismissRating) { Text(key: "action_cancel") }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    /// The wireframe's avatar and the two lines beside it.
    ///
    /// The photo is the placeholder glyph for ``DriverProfileScreen``'s reason: nothing in this target
    /// loads a remote image.
    private func passenger(_ rating: RatePassengerState) -> some View {
        HStack(spacing: MageRideSpacing.xs) {
            Image(systemName: "person.crop.circle.fill")
                .font(.system(size: MageRideControl.avatarSmall))
                .foregroundStyle(MageRideColor.onSurfaceVariant)

            VStack(alignment: .leading, spacing: 1) {
                Text(rating.passengerName ?? "history_passenger_unnamed".localised)
                    .mageFont(.title)
                    .foregroundStyle(MageRideColor.onSurface)
                Text(
                    rating.trip.routeText
                        + MageRideSymbols.separator
                        + ScheduleLabels.date(rating.trip.summary.startedAt)
                )
                .mageFont(.label)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
            }

            Spacer(minLength: 0)

            if rating.isLoading {
                ProgressView()
            }
        }
        .accessibilityElement(children: .combine)
    }
}

/// The five stars, as five tap targets.
///
/// `.isSelected` on the chosen one and an `accessibilityValue` on the row: one of five is exactly what a
/// star rating is, and VoiceOver announcing five buttons would leave a driver with no way to know which
/// was chosen. Each target is the HIG minimum, because a mis-tap here writes the wrong rating on
/// somebody's account.
private struct StarRow: View {

    let stars: Int
    let onStarsChange: (Int) -> Void

    var body: some View {
        HStack(spacing: MageRideSpacing.xs) {
            Spacer(minLength: 0)
            ForEach(RatePassengerState.minStars...RatePassengerState.maxStars, id: \.self) { star in
                Button { onStarsChange(star) } label: {
                    Text(star <= stars ? MageRideSymbols.starFilled : MageRideSymbols.starEmpty)
                        .mageFont(.headline)
                        .foregroundStyle(MageRideColor.warning)
                        .frame(
                            minWidth: MageRideControl.minimumTapTarget,
                            minHeight: MageRideControl.minimumTapTarget
                        )
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityLabel(Text("history_rate_stars".localisedFormat(star)))
                .accessibilityAddTraits(star == stars ? [.isButton, .isSelected] : .isButton)
            }
            Spacer(minLength: 0)
        }
        .accessibilityValue(Text("history_rate_stars".localisedFormat(stars)))
    }
}
